### PR TITLE
feat(comment): allow to choose between 3 comment strategies

### PR DESCRIPTION
**Pull request by JordanGoasdoue**
This PR allows to choose between 3 different comment strategies:
- create (to always create a comment, this is the default strategy and the one used before)
- create-if-not-exists (to create a comment only if its content doesn't already exist on the associated PR)
- delete-and-create (to delete comment(s) if the content is similar and create the comment again

The check / deletion is only done to comment with similar Author Login :ok_hand:  

Examples:
Create Strategy:
```
jx-gitops pr comment -s create -c ':star:TEST COMMENT:star:'
commented on pull request #22 on repository xxx/test-comment
```
Create If Not Exists Strategy
```
jx-gitops pr comment -s create-if-not-exists -c ':star:TEST COMMENT:star:'
Similar comment already exists on pull request #22 on repository xxx/test-comment
```
Delete and Create Strategy
```
jx-gitops pr comment -s delete-and-create -c ':star:TEST COMMENT:star:'
deleted comment with ID 12345 on pull request #22 on repository xxx/test-comment
commented on pull request #22 on repository xxx/test-comment
```

No defined Strategies (using the default one create)
```
jx-gitops pr comment -c ':star:TEST COMMENT:star:'
commented on pull request #22 on repository xxx/test-comment
```

#############

Tests are passing locally

```
go test -v ./pkg/cmd/pr/comment
=== RUN   TestPullRequestCommentStrategies
=== RUN   TestPullRequestCommentStrategies/CreateCommentStrategy-Existing-DifferentUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for CreateCommentStrategy
=== RUN   TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-Existing-DifferentUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for CreateIfNotExistsCommentStrategy
=== RUN   TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-Existing-DifferentUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for DeleteAndCreateCommentStrategy
=== RUN   TestPullRequestCommentStrategies/CreateCommentStrategy-Existing-SimilarUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for CreateCommentStrategy
=== RUN   TestPullRequestCommentStrategies/CreateCommentStrategy-NotExisting-SimilarUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: new-comment for CreateCommentStrategy
=== RUN   TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-Existing-SimilarUser
Similar comment already exists on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: existing-comment for CreateIfNotExistsCommentStrategy
=== RUN   TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-NotExisting-SimilarUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: new-comment for CreateIfNotExistsCommentStrategy
=== RUN   TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-Existing-SimilarUser
deleted comment with ID 0 on pull request #123 on repository myorg/myrepo
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for DeleteAndCreateCommentStrategy
=== RUN   TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-NotExisting-SimilarUser
commented on pull request #123 on repository myorg/myrepo
    comment_test.go:162: pull request #123, has comments: comment for DeleteAndCreateCommentStrategy
--- PASS: TestPullRequestCommentStrategies (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateCommentStrategy-Existing-DifferentUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-Existing-DifferentUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-Existing-DifferentUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateCommentStrategy-Existing-SimilarUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateCommentStrategy-NotExisting-SimilarUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-Existing-SimilarUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/CreateIfNotExistsCommentStrategy-NotExisting-SimilarUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-Existing-SimilarUser (0.00s)
    --- PASS: TestPullRequestCommentStrategies/DeleteAndCreateCommentStrategy-NotExisting-SimilarUser (0.00s)
PASS
ok      github.com/jenkins-x-plugins/jx-gitops/pkg/cmd/pr/comment       0.030s
```

### CODE REVIEW

File: pkg/cmd/pr/comment/comment.go [hunk 1]
```diff
@@ -16,6 +17,15 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const (
+	// Create a new comment even if it already exist
+	CreateCommentStrategy = "create"
+	// Create a new comment only if it doesn't already exist
+	CreateIfNotExistCommentStrategy = "create-if-not-exist"
```
Comment (vbehar) [inline pkg/cmd/pr/comment/comment.go:24]:
```suggestion
	CreateIfNotExistsCommentStrategy = "create-if-not-exists"
```

File: pkg/cmd/pr/comment/comment.go [hunk 2]
```diff
@@ -69,21 +86,93 @@ func (o *Options) Run() error {
...
+	existingComments, err := o.list(ctx, pr)
+	if err != nil {
+		return err
+	}
+
+	similarComments := make([]*scm.Comment, 0)
+	for _, comment := range existingComments {
+		if comment.Body == o.Comment {
+			similarComments = append(similarComments, comment)
+		}
```
Comment (vbehar) [inline pkg/cmd/pr/comment/comment.go:168]:
```suggestion
	for i := range existingComments {
	        comment := existingComments[i]
		if comment.Body == o.Comment {
			similarComments = append(similarComments, comment)
		}
```

until go 1.22 and the "for loop fix" - see https://tip.golang.org/doc/go1.22 - you'll have to take care of the fact that the `comment` variable is shared

Comment (JordanGoasdoue) [reply pkg/cmd/pr/comment/comment.go]: thank you for the review :) !
fixed

**Review by vbehar — APPROVED** [review]

**Review by msvticket — APPROVED** [review]
The code looks good

**Comment by joshuasimon-taulia** [comment]
can we get a "comment and hide previous" option similar how to [Atlantis](https://github.com/runatlantis/atlantis) handles PR comments? 
image

**Comment by JordanGoasdoue** [comment]
@joshuasimon-taulia didn't see your comment before but indeed you can if you want open a new PR with a new option to hide the previous comment :ok_hand:

**Pull request merged**